### PR TITLE
End-to-end tagging: C++

### PR DESCRIPTION
**Pull request by teh-cmc**
Implemented with the help of @Wumpf.

This semantically mimics very closely the way things are done in Rust, minus all technical differences due to the differences between both the languages and the SDKs.
For that reason, everything stated in https://github.com/rerun-io/rerun/pull/8304#issue-2715466746 basically applies as-is.

Pretty happy about it, I must say.

* DNM: requires #8304 
* Part of #7948

### CODE REVIEW

File: rerun_cpp/src/rerun/component_descriptor.hpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        /// Sets `archetype_field_name` to the given one iff it's not already set.
+        ComponentDescriptor or_with_archetype_field_name(const char* archetype_field_name_) const {
+            if (this->archetype_field_name.has_value()) {
+                return *this;
+            }
+            ComponentDescriptor descriptor = *this;
+            descriptor.archetype_field_name = archetype_field_name_;
+            return descriptor;
+        }
```
Comment (teh-cmc) [inline rerun_cpp/src/rerun/component_descriptor.hpp:127]: Note: I tried to not repeat myself so much, but anytime I try to do something any fancier than that, either clang and/or gcc complains or things just segfault. Whatever.

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
TODO: end-to-end failed on arm64 mac:
```
 FAILED: docs/snippets/CMakeFiles/descr_custom_archetype.dir/all/descriptors/descr_custom_archetype.cpp.o 
/Users/runner/work/rerun/rerun/.pixi/envs/cpp/bin/x86_64-apple-darwin13.4.0-clang++  -I/Users/runner/work/rerun/rerun/rerun_cpp/src -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/runner/work/rerun/rerun/.pixi/envs/cpp/include -g -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wall -Wcast-align -Wcast-qual -Wextra -Wformat=2 -Wmissing-include-dirs -Wnull-dereference -Wold-style-cast -Wpedantic -Wpointer-arith -Wshadow -Wswitch-enum -Wunreachable-code -Wvla -Wc++17-compat-pedantic -Wc++20-compat-pedantic -Wc99-extensions -Weverything -Wgnu -Wnon-gcc -Wpre-c2x-compat-pedantic -Wshadow-all -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-covered-switch-default -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-padded -Wno-reserved-id-macro -Wno-reserved-identifier -Wno-unreachable-code-break -Wno-unreachable-code-return -Wno-unused-macros -Wno-unsafe-buffer-usage -Wno-unknown-warning-option -Werror -g -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -Werror -MD -MT docs/snippets/CMakeFiles/descr_custom_archetype.dir/all/descriptors/descr_custom_archetype.cpp.o -MF docs/snippets/CMakeFiles/descr_custom_archetype.dir/all/descriptors/descr_custom_archetype.cpp.o.d -o docs/snippets/CMakeFiles/descr_custom_archetype.dir/all/descriptors/descr_custom_archetype.cpp.o -c /Users/runner/work/rerun/rerun/docs/snippets/all/descriptors/descr_custom_archetype.cpp
/Users/runner/work/rerun/rerun/docs/snippets/all/descriptors/descr_custom_archetype.cpp:68:38: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
    CustomPosition3D positions[1] = {rerun::components::Position3D{1.0f, 2.0f, 3.0f}};
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     {                                              }
1 error generated.
[597/617] Building CXX object docs/snippets/CMakeFiles/descr_builtin_component.dir/all/descriptors/descr_builtin_component.cpp.o
[598/617] Building CXX object docs/snippets/CMakeFiles/different_data_per_timeline.dir/all/concepts/different_data_per_timeline.cpp.o
[599/617] Building CXX object docs/snippets/CMakeFiles/descr_builtin_archetype.dir/all/descriptors/descr_builtin_archetype.cpp.o
[600/617] Linking CXX executable docs/snippets/native-async
[601/617] Building CXX object docs/snippets/CMakeFiles/descr_custom_component.dir/all/descriptors/descr_custom_component.cpp.o
FAILED: docs/snippets/CMakeFiles/descr_custom_component.dir/all/descriptors/descr_custom_component.cpp.o 
/Users/runner/work/rerun/rerun/.pixi/envs/cpp/bin/x86_64-apple-darwin13.4.0-clang++  -I/Users/runner/work/rerun/rerun/rerun_cpp/src -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/runner/work/rerun/rerun/.pixi/envs/cpp/include -g -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wall -Wcast-align -Wcast-qual -Wextra -Wformat=2 -Wmissing-include-dirs -Wnull-dereference -Wold-style-cast -Wpedantic -Wpointer-arith -Wshadow -Wswitch-enum -Wunreachable-code -Wvla -Wc++17-compat-pedantic -Wc++20-compat-pedantic -Wc99-extensions -Weverything -Wgnu -Wnon-gcc -Wpre-c2x-compat-pedantic -Wshadow-all -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-covered-switch-default -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-padded -Wno-reserved-id-macro -Wno-reserved-identifier -Wno-unreachable-code-break -Wno-unreachable-code-return -Wno-unused-macros -Wno-unsafe-buffer-usage -Wno-unknown-warning-option -Werror -g -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -Werror -MD -MT docs/snippets/CMakeFiles/descr_custom_component.dir/all/descriptors/descr_custom_component.cpp.o -MF docs/snippets/CMakeFiles/descr_custom_component.dir/all/descriptors/descr_custom_component.cpp.o.d -o docs/snippets/CMakeFiles/descr_custom_component.dir/all/descriptors/descr_custom_component.cpp.o -c /Users/runner/work/rerun/rerun/docs/snippets/all/descriptors/descr_custom_component.cpp
/Users/runner/work/rerun/rerun/docs/snippets/all/descriptors/descr_custom_component.cpp:32:38: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
    CustomPosition3D positions[1] = {rerun::components::Position3D{1.0f, 2.0f, 3.0f}};
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     {                                              }
1 error generated.
```

**Comment by teh-cmc** [comment]
@rerun-bot full-check

File: crates/build/re_types_builder/src/codegen/cpp/mod.rs [hunk 2]
```diff
@@ -1576,6 +1576,7 @@ fn to_arrow_method(
 
 fn archetype_serialize(type_ident: &Ident, obj: &Object, hpp_includes: &mut Includes) -> Method {
     hpp_includes.insert_rerun("component_batch.hpp");
+    hpp_includes.insert_rerun("component_descriptor.hpp");
```
Comment (Wumpf) [inline crates/build/re_types_builder/src/codegen/cpp/mod.rs:1579]: we actually don't use `ComponentDescriptor` in the header of the archetype at all.
We do use it in the cpp, but we already include the types for concrete components there anyways, so we know about `ComponentDescriptor` transitively

```suggestion
```

**Review by Wumpf — APPROVED** [review]
looking great!
Didn't understand though how indicator components work now in this world - did you just change the interim definitions of indicator components to not contain the archetype name?

**Comment by teh-cmc** [comment]
> Didn't understand though how indicator components work now in this world - did you just change the interim definitions of indicator components to not contain the archetype name?

For now, indicators just set their component_name (same is true of all languages, so that roundtrips pass).

This will change soon when we enter milestone 2: https://github.com/rerun-io/rerun/issues/8129#issuecomment-2519783103, but one issue at a time.

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
oh god what now :joy:

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
this is the one

**Pull request merged**